taf/testlib/virtual_env.py

```diff
@@ -291,22 +292,37 @@ def ensure_public_access(self, try_reuse=True, networks_client=None, routers_cli
 
         # Try to reuse existing stuff that meets requirements, if desirable (devstack)
         if try_reuse and self._reuse_public_access(net_ip, routers_client=routers_client, tenant_id=tenant_id):
-            self.class_logger.debug('Reused')
+            self.class_logger.debug('public net and router were Reused')
             return True
 
-        self.class_logger.debug('No reuse')
-        # create new public router & network
-        public_network_kwargs = {
-            'routers_client': routers_client,
-            'networks_client': networks_client,
-            'delete_external': True,
-            'name': name,
-            'tenant_id': tenant_id,
-            'external_net': net_ip,
-        }
-        public_network = self.create_public_network(**public_network_kwargs)
-        assert public_network
+        public_network = None
+        # Try to reuse public network created by different tenant (needed for magnum)
+        if try_reuse:
+            public_network = self._reuse_public_network(net_ip, routers_client=routers_client, networks_client=networks_client, tenant_id=tenant_id)
+        
+        if public_network:
+            self.class_logger.debug('public net was Reused')
+        else:
+            self.class_logger.debug('No reuse')
+
+            del_external_kwargs = {
+                'routers_client': routers_client,
+                'networks_client': networks_client,
+                'ports_client': ports_client,
+                'tenant_id': tenant_id,
+            }
+            external_net_ids = self._delete_external_elements(**del_external_kwargs)
 
+            # create new public network
+            public_network_kwargs = {
+                'networks_client': networks_client,
+                'name': name,
+                'tenant_id': tenant_id,
+            }
+            public_network = self.create_public_network(**public_network_kwargs)
+            assert public_network
```

Comment:
please avoid asserting something that will be more tested by an upcoming statement.  i hope we are not using AssertionErrors as exceptions to recover from.